### PR TITLE
Enable delayed type annotation evaluation according to PEP 563

### PR DESCRIPTION
**Pull request by justinchuby**
- Add `from __future__ import annotations` to all files needed
- Use the new style type annotation https://peps.python.org/pep-0604/
- Fix mypy errors

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -2,15 +2,16 @@
 # Please move this into pyproject.toml, when it does.
 
 [flake8]
-max-line-length = 120  # 120 is a hard limit. Black keeps lines generally at 88.
+# 120 is a hard limit. Black keeps lines generally at 88.
```
Comment (justinchuby) [inline .flake8:5]: Fix syntax

File: .flake8 [hunk 2]
```diff
@@ -2,15 +2,16 @@
...
+max-line-length = 120
 
 # W503 talks about operator formatting which is too opinionated.
 # E203 is need to support black formatting.
 # https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html
 # Dxx ignores are for pydocstyle. It should mirror the [tool.pydocstyle] section in
 # pyproject.toml. Mirrored here for flake8-docstring support. See pyproject.toml for
 # reasons for disabling.
-extend-ignore = E203, W503, D1, D202, D205, D041, D415
```
Comment (justinchuby) [inline .flake8:13]: typo

**Comment by gramalingam** [comment]
Out of curiosity: is there a compelling reason to replace `Optional[T]` by `T | None` ? It seems to me that the first is a more readable abbreviation. Replacing `Union(X, Y)` by `X | Y` makes sense.

**Review by gramalingam — APPROVED** [review]

**Comment by justinchuby** [comment]
Done

**Pull request merged**